### ISSUE TITLE
IntSet range construction

### ISSUE THREAD

**Issue by treeowl** [haskell/containers#632]
`fromDistinctAscList [a,b..c]` takes time proportional to the number of elements. We should be able to write a specialized function that does better when `b-a` is less than 58 or so, and significantly better when it is very small. The idea is to fill up a whole `Tip` in one go.

**Comment by mikeplus64** [comment]
It might be nice to just have a special `enumFromThenTo` for `IntSet`. `fromDistinctAscList [a, b .. c]` can then use a rewrite rule to use that.

**Comment by treeowl** [comment]
@mikeplus64 , such a rewrite rule would probably be impossible to write. `enumFromThenTo` gets rewritten to use `build`, and I don't think the things it's rewritten to are exported. In any case, it's a godawful mess and we should probably steer clear. But yes, a specialized version of some sort seems quite reasonable.

**Comment by mikeplus64** [comment]
Ah, that's a shame. But yeah `enumFromTo` and `enumFromThenTo` would still be nice and probably much simpler (i.e. more performant) due to knowing upfront the whole range.

**Comment by treeowl** [comment]
Indeed, it's likely somewhat helpful to know the final tree shape in advance.

**Comment by amigalemming** [comment]
I like to have that function in order to initialize a range as start set for a prime sieve.

**Comment by meooow25** [comment]
This should be pretty straightforward to implement. I can put together a PR.

**Comment by meooow25** [comment]
Somehow I managed to overlook that that original comment is about `[a,b..c]`, thinking that it's `[a..b]` instead. `[a,b..c]` won't be that straightforward. Anyway, no harm in adding the `[a..b]` version I think.

**Comment by meooow25** [comment]
Thinking about it again, the `[a,b..c]` version seems very specific and also uncommon. It just happens to be well known to us because of `enumFromThenTo`.

I can find quite a few instances of `IntSet.from.*List [a..b]` on Hackage, but none of `[a,b..c]`. Some examples:

* [Agda](https://hackage-search.serokell.io/viewfile/Agda-2.7.0.1/src/full/Agda/TypeChecking/Telescope.hs#line-328)
* [alfred-margaret](https://hackage-search.serokell.io/viewfile/alfred-margaret-2.1.0.2/src/Data/Text/AhoCorasick/Searcher.hs#line-176)
* [GHC](https://hackage-search.serokell.io/viewfile/ghc-9.12.1/GHC/Cmm/Dataflow.hs#line-211)
* [ormulo](https://hackage-search.serokell.io/viewfile/ormolu-0.8.0.0/src/Ormolu/Diff/Text.hs#line-106)

We have `fromRange`, so I'm closing this. Please reopen if someone really wants the `[a,b..c]` version.
